rfcs/syntax-labeled-break-continue.md

```diff
@@ -0,0 +1,54 @@
+# Labeled Break and Continue
+
+## Summary
+
+Labeled break and continue statements allow for specifying which loop is being broken out of or continued into in the case of multiple nested loops. 
+
+## Motivation
+
+Luau does not have a clean way to deal with exiting multiple nested loops, or advance an outermost loop while skipping execution in all the inside ones.
+```Lua
+while cond_1 do
+    -- code
+    while cond_2 do
+        -- code
+        break -- we can not break the 2nd loop from here
+    end
+end
+```
+This kind of operation is useful in a lot of places, although not very common. Another interesting application of it would be facilitating code generation with tooling that supports this behavior already.
+
+## Design
+
+The proposed implementation is to allow for vanilla Lua-like labels, with the `::name::` style, but only before loop statements. This is essentially a `goto` statement that can only go outwards and not to an arbitrary location. The behavior is not much different than WebAssembly's branching instructions, or Rust's `break 'label`/`continue 'label` syntax. A possible implementation might look like the following code:
+```Lua
+::my_loop:: while true do
+    while true do
+        break my_loop
+    end
+end
+```
+
+Though, this example is ambiguous on `expr :: type` syntax and for any already invalid statements that could come after `break`. A possible solution might be using another symbol, possibly a new unused one or recycling an operator.
```

Comment:
`my_loop: while true do` should be completely unambiguous, since `while` is not a valid type name, and you can't just start with an identifier.

Reply:
Requires some amount of backtracking or AST inspection to parse because normally `myloop:while` is a function call.

Reply:
You could take a page from pratt parsing and carry forward `label_name`. Then do one token lookahead (which we have) to see if `:` is followed by one of the token that starts a loop construct.

Reply:
Right, but if it isn't you will need to reparse `label_name` as global/local. Which is possible, it alters the parsing flow in a somewhat odd way but doable.

Reply:
I reckon there are no symbols we can recycle for the left hand side to avoid ambiguity? I also briefly considered the alternative of `@label` but that introduces a new symbol which I'm not sure is ideal.

I also considered disambiguating it similarly to how calls on different lines works in vanilla, where a syntax error is thrown when it's ambiguous.

If there are no better alternatives, then `label:` is an okay compromise.

Reply:
We plan to use `@` for attributes in the future

Reply:
It seems the biggest languages with the feature have the following syntax:
```Rust
// Rust
'label while cond { break 'label; }
```
```Kotlin
// Kotlin
label@ while (cond) { break @label }
```
```Java
// Java
label: while (cond) { break label; }
```
```Go
// Go
label: for cond { break label }
```
The Rust example is the most attractive because it requires no lookahead, but would be odd to look at in a language where `'` is only used for strings. The others are all kind of the same in that they have `label` followed by a discriminant symbol, which was already discussed to be possible but not optimal.

Of course it's also possible to discuss other forms of labeling, such as inner loop labeling:
```Lua
while cond do ::label:: end
```
This would completely disambiguate but may look odd for people that expect labels to be definitive positions where the code will jump to, while in this case a `continue label` for example would jump into the loop condition instead.

It could also be incorporated into the loop itself which could serve to fix both of these issues but still look a little odd:
```Lua
while ::label:: cond do end
```

Though I'm just throwing possibilities to see if any stick.